library/test/src/types.rs

```diff
@@ -112,9 +112,13 @@ impl fmt::Debug for TestFn {
     }
 }
 
+// A unique integer associated with each test.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct TestId(pub usize);
+
 // The definition of a single test. A test runner will run a list of
 // these.
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
```

Comment:
I guess an alternative here could be to add an `id` field to `TestDesc`. If all we need is to guarantee that different tests with the same description don't equate that should even be enough to leave the derived equality and hashing.

Reply:
If we added `id` to `TestDesc`, the responsibility for generating it and ensuring its uniqueness would fall to the compiler (I’m not sure if this macro even *can* generate a unique ID):

https://github.com/rust-lang/rust/blob/3e826bb11228508fbe749e594038d6727208aa94/compiler/rustc_builtin_macros/src/test.rs#L227-L297

as well as librustdoc:

https://github.com/rust-lang/rust/blob/3e826bb11228508fbe749e594038d6727208aa94/src/librustdoc/doctest.rs#L831-L842

and all the [tests for libtest itself](https://github.com/rust-lang/rust/blob/3e826bb11228508fbe749e594038d6727208aa94/library/test/src/tests.rs). Even if this is somehow possible, it wouldn’t reduce churn.

Another alternative would be to try using the *address* of the `TestDesc` as a unique identifier—but then instead of moving or cloning `TestDesc` we’d have to pass it by reference, which also at best wouldn’t reduce churn.